main/src/ca/uwaterloo/flix/language/phase/unification/set/Equation.scala

```diff
@@ -17,31 +17,47 @@
 package ca.uwaterloo.flix.language.phase.unification.set
 
 import ca.uwaterloo.flix.language.ast.SourceLocation
+import ca.uwaterloo.flix.language.phase.unification.set.SetFormula.*
 
 import scala.annotation.nowarn
 
 /**
   * Represents an equality equation between the formulas `f1` and `f2` (`f1 ~ f2`).
   *
+  * Equations can be marked as [[Equation.Status.Unsolvable]] during unification, in which case its
+  * written `f1 !~ f2`.
+  *
   * The `@nowarn` annotation is required for Scala 3 compatibility, since the derived `copy` method
   * is private in Scala 3 due to the private constructor. In Scala 2 the `copy` method is still
   * public. However, we do not use the `copy` method anywhere for [[Equation]], so this is fine.
   */
 @nowarn
-case class Equation private(f1: SetFormula, f2: SetFormula, loc: SourceLocation) {
+case class Equation private(f1: SetFormula, f2: SetFormula, status: Equation.Status, loc: SourceLocation) {
 
   /** Returns the sum of the sizes of the formulas in `this`. */
   final def size: Int = f1.size + f2.size
 
   /** Returns a human-readable string of `this`. */
   override final def toString: String = s"$f1 ~ $f2"
 
+  /** Returns a copy of `this` with the new `status` */
+  final def copyWithStatus(status: Equation.Status): Equation = {
+    if (status == this.status) this
+    else Equation(f1, f2, status, loc)
+  }
+
+  /** Returns a copy of `this` with status [[Equation.Status.Unsolvable]]. */
+  @inline
```

Comment:
I don't think these inline annotations should matter. The JVM is pretty much guaranteed to inline here.

Reply:
It is usefull though - I did not put `final` and the optimizer warned me that this could be overridden so It couldn't inline.

I'd guess it'd be jitted fine afterwards, but it shows that writing your intent is useful.